File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var distributeCoins = function(root,parent) {
    if(!root){
        return 0;
    }
    let moves=distributeCoins(root.left,root)+distributeCoins(root.right,root);
    let x=root.val-1;
    if(parent){
        parent.val+=x;
    }
    moves+=Math.abs(x);
    return moves
};

File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
 int dfs(struct TreeNode* root,struct TreeNode* parent){
    if(!root){
        return 0;
    }
    int moves=dfs(root->left,root)+ dfs(root->right,root);
    int x=root->val-1;
    if(parent){
        parent->val+=x;
    }
    moves+=abs(x);
    return moves;
 }
int distributeCoins(struct TreeNode* root) {
    return dfs(root,NULL);
    
}